pkg/rke2/control_plane.go

```diff
@@ -0,0 +1,343 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package rke2
+
+import (
+	"context"
+
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	"github.com/rancher-sandbox/cluster-api-provider-rke2/pkg/machinefilters"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apiserver/pkg/storage/names"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+
+	bootstrapv1 "github.com/rancher-sandbox/cluster-api-provider-rke2/bootstrap/api/v1alpha1"
+
+	controlplanev1 "github.com/rancher-sandbox/cluster-api-provider-rke2/controlplane/api/v1alpha1"
+	"sigs.k8s.io/cluster-api/controllers/external"
+
+	"sigs.k8s.io/cluster-api/util/patch"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// ControlPlane holds business logic around control planes.
+// It should never need to connect to a service, that responsibility lies outside of this struct.
+// Going forward we should be trying to add more logic to here and reduce the amount of logic in the reconciler.
+type ControlPlane struct {
+	RCP                  *controlplanev1.RKE2ControlPlane
+	Cluster              *clusterv1.Cluster
+	Machines             FilterableMachineCollection
+	machinesPatchHelpers map[string]*patch.Helper
+
+	// reconciliationTime is the time of the current reconciliation, and should be used for all "now" calculations
+	reconciliationTime metav1.Time
+
+	// TODO: we should see if we can combine these with the Machine objects so we don't have all these separate lookups
+	// See discussion on https://github.com/kubernetes-sigs/cluster-api/pull/3405
+	rke2Configs    map[string]*bootstrapv1.RKE2Config
+	infraResources map[string]*unstructured.Unstructured
+}
+
+// NewControlPlane returns an instantiated ControlPlane.
+func NewControlPlane(ctx context.Context, client client.Client, cluster *clusterv1.Cluster, kcp *controlplanev1.RKE2ControlPlane, ownedMachines FilterableMachineCollection) (*ControlPlane, error) {
```

Comment:
should this be located in the `controlplane` package? are we going to use this method anywhere outside `controlplane`?

Reply:
No it is not used anywhere else.